### PR TITLE
Reduce dependency of `BalanceSpec` on `cardano-wallet`.

### PR DESCRIPTION
**Pull request by jonathanknowles**
## Issue

ADP-3171

## Description

This PR reduces the degree to which `Write.Tx.BalanceSpec` depends on the `cardano-wallet` library.

We achieve this by:
- conservatively inlining and specialising a small number of functions.
- moving the various `maxLengthAddressFor` constants to the `address-derivation-discovery` library.

## Motivation

Make it possible to move the `BalanceSpec` module to the `cardano-balance-tx` library.

### CODE REVIEW

File: lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs [hunk 1]
```diff
@@ -1973,22 +1978,25 @@ balanceTx
...
-                protocolParameters
-                timeTranslation
-                utxoAssumptions
-                utxoIndex
-                genChange
-                s
-                partialTx
-        pure transactionInEra
+    =
+    case changeAddrGen of
```
Comment (Anviking) [inline lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs:1987]: IMO we can simply use multiple lines for the pattern match: https://github.com/cardano-foundation/cardano-wallet/commit/4fafaf646e28475f3e5bc974bdcd00b82c01c8b2

Although if we drop the `'` from `Wallet` we do have enough room to pattern match it in a single line:

https://github.com/cardano-foundation/cardano-wallet/commit/e90747020a7f60a39ce1ba2af388fb05104b6eef

https://github.com/cardano-foundation/cardano-wallet/commit/a6995b637f6c5be4eef8f84102f19c18d979eeb7

(feel free to cherry-pick)

**Review by Anviking — APPROVED** [review]
Fair enough, this approach seems to work - nice!

File: lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs:1987]: Hi @Anviking 

Many thanks for these suggestions -- I'll try to cherry-pick one of them in a follow-up PR.

Comment (jonathanknowles) [reply lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs:1987]: @Anviking I've included c060bbcdaa76a6aa8a3270707a45d5a987289905 within #4313.

**Pull request merged**